File: auth-lambda/jwt_service.py
import jwt
import datetime


class JWTService:
    def __init__(self, secret: str, algorithm: str = "HS256", expiry_hours: int = 24):
        self.secret = secret
        self.algorithm = algorithm
        self.expiry_hours = expiry_hours

    def generate(self, user_id: str, email: str, role: str):
        now = datetime.datetime.utcnow()
        expires_at = now + datetime.timedelta(hours=self.expiry_hours)
        payload = {
            "sub": user_id,
            "email": email,
            "role": role,
            "iat": now,
            "exp": expires_at,
        }
        token = jwt.encode(payload, self.secret, algorithm=self.algorithm)
        return token, expires_at.isoformat() + "Z"

    def validate(self, token: str) -> dict:
        return jwt.decode(token, self.secret, algorithms=[self.algorithm])

File: internal/infrastructure/auth/jwt_service.go
package auth

import (
	"errors"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

type JWTClaims struct {
	jwt.RegisteredClaims
	Email string   `json:"email"`
	Roles []string `json:"roles"`
}

type JWTService struct {
	secret      string
	expiryHours int
}

func NewJWTService(secret string, expiryHours int) *JWTService {
	return &JWTService{secret: secret, expiryHours: expiryHours}
}

func (s *JWTService) Generate(userID, email string, roles []string) (string, time.Time, error) {
	now := time.Now()
	expiresAt := now.Add(time.Duration(s.expiryHours) * time.Hour)
	claims := JWTClaims{
		RegisteredClaims: jwt.RegisteredClaims{
			Subject:   userID,
			IssuedAt:  jwt.NewNumericDate(now),
			ExpiresAt: jwt.NewNumericDate(expiresAt),
		},
		Email: email,
		Roles: roles,
	}
	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	signed, err := token.SignedString([]byte(s.secret))
	if err != nil {
		return "", time.Time{}, err
	}
	return signed, expiresAt, nil
}

func (s *JWTService) Validate(tokenStr string) (*JWTClaims, error) {
	token, err := jwt.ParseWithClaims(tokenStr, &JWTClaims{}, func(t *jwt.Token) (interface{}, error) {
		if _, ok := t.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, errors.New("unexpected signing method")
		}
		return []byte(s.secret), nil
	})
	if err != nil {
		return nil, err
	}
	claims, ok := token.Claims.(*JWTClaims)
	if !ok || !token.Valid {
		return nil, errors.New("invalid token")
	}
	return claims, nil
}
